rfc/0009-policy-report.md

```diff
@@ -0,0 +1,323 @@
+
+* How many policy evaluations have been done during the audit run of the Namespace
+* How many policy evaluations led to a resource being accepted, rejected or
+  could not be performed because of an error.
+  Policy failures happen when an error occurs during the policy evaluation,
+  for example, because of a runtime panic
+* Detailed results about all the evaluations performed. An evaluation result contains
+  the following information: the policy being used, the resource that was
+  evaluated, the final outcome (accept/reject/failure) and the eventual message
+  returned by the policy
+
+The `PolicyReport` of the Namespace is kept updated during the lifetime of the
+cluster, meaning:
+
+* New Policies might become "interested" about the events happening inside of
+  the cluster. New evaluations results will be added to the PolicyReport
+* Existing Kubernetes objects and policies can change over time: some evaluations
+  could change their outcome (pass -> fail and vice versa). The evaluation results
+  must reflect that
+* Policies might be deleted/might be no longer "interested" about the Namespace:
+  the report must be updated to not container any reference the these policies
+  and their evaluation results
+* Kubernetes resources might be deleted: the report must be updated to not contain
+  any reference to them
+
+## Cluster report
+
+All the Kubernetes resources that are not namespaced, are going to be audited
+separately.
+The auditing results are going to be saved into a Custom Resource called:
+`ClusterPolicyReport`.
+
+Given `ClusterPolicyReport` has the same structure as `PolicyReport`, Kubewarden
+will handle it in the same way described in the previous section.
+
+## `PolicyReport` detailed overview
+
+For each Namespace, Kubewarden will find all the policies that are interested
+about actions happening inside of it. This is going to be a mix of
+some `ClusterAdmissionPolicy` policies and all the `AdmissionPolicy` policies
+that are defined inside of the Namespace.
+
+Next, this list of policies is going to be filtered: only policies interested
+in `CREATE` events are going to be considered.
+The audit report is not going to simulate `DELETE` and `UPDATE` events.
+
+For each relevant policy, Kubewarden will find all the Kubernetes resources that
+are 1) defined inside of the Namespace and 2) are relevant to the policy. Kubewarden
+will then simulate the `CREATE` event of this resource and keep track of the policy
+evaluation result.
+
+All the results are going to be collected into the `PolicyReport` object defined
+inside of the Namespace. There's going to be only one `PolicyReport` per Namespace.
+
+The `PolicyReport` resource has the following fields:
+
+* `apiVersion` and `kind`: these are defined by the CRDs
+* `metadata`: this is the usual `meta/v1.ObjectMeta` resource. In the
+  beginning we will set only the `name` field. We will do that by using the
+  following pattern: `polr-ns-<namespace name>`
```

Comment:
I wonder if we should add versioning to our specific usage of PolicyReports. If so, maybe a `metadata.annotation.version == v1`?

Reply:
Do you think we should do that because we might change the way we "consume" the `PolicyReport` document in the future?

Reply:
More that I think we may extend our usage by adding `PolicyReportResult` `properties` and so on. Both have admittedly low probabilities.
I'm happy without versioning them, don't intend to block on it.